tests/test_models.py

```diff
@@ -0,0 +1,237 @@
+        parent.refresh_from_db()
+        assert parent.state == 'Custom State'
+
+    def test_start(self):
+        """The start() method should update the status record appropriately."""
+        status = self._status(state=UserTaskStatus.PENDING)
+        assert status.state == UserTaskStatus.PENDING
+        status.start()
+        assert status.state == UserTaskStatus.IN_PROGRESS
+        status.refresh_from_db()
+        assert status.state == UserTaskStatus.IN_PROGRESS
+
+    def test_start_canceled_task(self):
+        """Attempting to start a canceled task should raise a TaskCanceledException."""
+        status = self._status(state=UserTaskStatus.CANCELED)
+        with pytest.raises(TaskCanceledException):
+            status.start()
+        assert status.state == UserTaskStatus.CANCELED
+
+    def test_start_with_failed_sibling(self):
+        """If a sibling task has already failed, the parent status should not change when the task starts."""
+        parent = self._status(is_container=True, state=UserTaskStatus.FAILED)
+        child1 = self._status(parent=parent, state=UserTaskStatus.FAILED)
+        child2 = self._status(parent=parent, state=UserTaskStatus.PENDING)
+        child2.start()
+        parent.refresh_from_db
+        assert child2.state == UserTaskStatus.IN_PROGRESS
+        assert parent.state == UserTaskStatus.FAILED
+
+    def test_start_with_parent(self):
+        """Starting a pending task should also update any parent task status."""
+        parent = self._status(is_container=True, state=UserTaskStatus.PENDING)
+        status = self._status(parent=parent, state=UserTaskStatus.PENDING)
+        status.start()
+        assert status.state == UserTaskStatus.IN_PROGRESS
+        parent.refresh_from_db()
+        assert parent.state == UserTaskStatus.IN_PROGRESS
+
+    def test_string_representation(self):
+        """UserTaskStatus instances should have a reasonable string representation."""
+        status = self._status()
+        assert str(status) == '<UserTaskStatus: SampleTask>'
+
+    def test_succeed(self):
+        """succeed() should update its parent only if all its other children have already succeeded."""
+        parent = self._status(is_container=True, total_steps=10)
+        child1 = self._status(parent=parent)
+        child2 = self._status(parent=parent)
+        child1.succeed()
+        parent.refresh_from_db()
+        assert child1.state == UserTaskStatus.SUCCEEDED
+        assert child1.completed_steps == child1.total_steps
+        assert parent.state == UserTaskStatus.IN_PROGRESS
+        assert parent.completed_steps == child1.total_steps
+        child2.succeed()
+        parent.refresh_from_db()
+        assert child2.state == UserTaskStatus.SUCCEEDED
+        assert child2.completed_steps == child2.total_steps
+        assert parent.state == UserTaskStatus.SUCCEEDED
+        assert parent.completed_steps == child1.total_steps + child2.total_steps
+
+    def _status(self, **kwargs):
```

Comment:
When I'm reading code, I dislike running across a function and not knowing where it's defined. I tend to define the code that uses the utility code last. Modern debuggers/editors make finding these methods effortless, but I've never claimed to be modern. Bottom line: It's all good.